### PR TITLE
Added IsEmpty operation for filter predicate

### LINKED ISSUE #1160: Add IsEmpty Filter Operator
Elide should provide a filter operator to test if a collection is empty.

## Expected Behavior

Support filters of the variety.

`filter=book.authors=isempty=true`
`filter=book.authors=isempty=1`

Both attribute and relationship collections should be supported.

## Current Behavior
Filters that span relationships is partially supported in JSON-API and not supported in GraphQL.  Issue #1141 will remedy that.

Also, filters today are only supported on model attributes.  This needs to be expanded to include relationships in limited operator contexts (like this one).

## Possible Solution
We must first address #1141 

This operator will likely require different behavior for JPQL stores and for in memory filtering.  

For JPQL stores, we'll need to use the collection empty operator:
https://docs.oracle.com/html/E13946_04/ejb3_langref.html#ejb3_langref_empty_comp

For in memory stores, we'll need to check if the field is null OR if the collection is empty.

### PR DESCRIPTION
**Pull request by Chandrasekar-Rajasekar**
Resolves #1160 

## Description
Added IsEmpty Filter operation with syntax `=isempty=`. 

## Motivation and Context
isempty filter operation are useful in most cases and the community needed support for this operation. 

## How Has This Been Tested?
Added New Test Cases

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-core/src/main/java/com/yahoo/elide/core/filter/dialect/RSQLFilterDialect.java [hunk 1]
```diff
@@ -368,5 +377,25 @@ private FilterExpression buildIsNullOperator(Path path, List<String> arguments)
...
+        private FilterExpression buildIsEmptyOperator(Path path, List<String> arguments) {
+            String arg = arguments.get(0);
+            try {
+                boolean wantsEmpty = CoerceUtil.coerce(arg, boolean.class);
+                if (wantsEmpty) {
+                    return new IsEmptyPredicate(path);
+                }
+                return new NotEmptyPredicate(path);
+            } catch (InvalidValueException ignored) {
+                throw new RSQLParseException(String.format("Invalid value for operator =isnull= '%s'", arg));
```
Comment (PatRolland) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/dialect/RSQLFilterDialect.java]: copy pasted =isnull=. here should be =isempty=

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java [hunk 2]
```diff
@@ -316,6 +334,24 @@ public static Operator fromString(final String string) {
         return (T entity) -> false;
     }
 
+    private static <T> Predicate<T> isEmpty(String field, RequestScope requestScope) {
+        return (T entity) -> {
+            Object val = getFieldValue(entity, field, requestScope);
+            if (val == null) { return true; }
+            if (val instanceof String) {
```
Comment (PatRolland) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: i feel like only collection and map should be handled here

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java [hunk 3]
```diff
@@ -316,6 +334,24 @@ public static Operator fromString(final String string) {
         return (T entity) -> false;
     }
 
+    private static <T> Predicate<T> isEmpty(String field, RequestScope requestScope) {
+        return (T entity) -> {
+            Object val = getFieldValue(entity, field, requestScope);
+            if (val == null) { return true; }
```
Comment (PatRolland) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: null and empty do not represent the same state imo

Comment (Chandrasekar-Rajasekar) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: `For in memory stores, we'll need to check if the field is null OR if the collection is empty.` This is the requirement given in the issue.

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java (hunk 2 above)
Comment (Chandrasekar-Rajasekar) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: So the string `""` is not considered empty?

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java (hunk 3 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: I agree - they are not the same thing.  But I think you could interpret this either way.

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java (hunk 2 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: So we generate JPQL queries based on IS EMPTY and IS NOT EMPTY.  I'm not sure those will work on Strings when we push this to the datastore.  

I'm inclined to think the JPQL queries will blow up if we try this on Strings in the DB.  Either remove the String case - or we need to add explicit tests to see that they work as expected.

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java (hunk 3 above)
Comment (PatRolland) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: if you look at it in a db table way, a column referring to children might be null or have a value with no associations .
consider 3 tables
TableA(a_id), TableB(b_id) and TableAB(ab_id, a_id, b_id).
You might end up with a row in TableAB such as ab_id = 1, a_id = 1 and b_id = null thus representing an empty collection.
You migth also end up having no row in TableAB thus representing a null collection.

I'm not sure if this is a concept worth representing in the =isempty= filter but that is how I can represent the distinction here.

Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: @jkusa Any thoughts on this one?

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java (hunk 2 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: @Chandrasekar-Rajasekar Let's remove the string case.

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java (hunk 3 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: We'll run a test and see what JPQL does.  The in-memory result should match the JPQL result.

File: elide-datastore/elide-datastore-hibernate5/pom.xml [hunk 4]
```diff
@@ -37,6 +37,10 @@
         <url>https://github.com/yahoo/elide.git</url>
         <tag>HEAD</tag>
     </scm>
+    <properties>
+        <excludeTags>emptyOnAttributeCollection</excludeTags>
```
Comment (Chandrasekar-Rajasekar) [inline elide-datastore/elide-datastore-hibernate5/pom.xml]: Looks like the fix for querying attribute collection on EMPTY operator is fixed in version `5.4.11`. We are using `5.4.2.Final`. 
One option is we use `where size(book.awards) = 0` as suggested in below link. 
Or we could continue using Empty operator and change the version to 5.4.11 when it gets released.

https://hibernate.atlassian.net/browse/HHH-6686?attachmentOrder=asc&oldIssueView=true
http://hibernate.org/orm/releases/5.4/

Comment (aklish) [reply elide-datastore/elide-datastore-hibernate5/pom.xml]: We can update Hibernate version on this branch as well.

Comment (Chandrasekar-Rajasekar) [reply elide-datastore/elide-datastore-hibernate5/pom.xml]: It is not released yet http://hibernate.org/orm/releases/5.4/

File: elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java (hunk 3 above)
Comment (PatRolland) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: Makes sense. Is there a way where I could make the distinction between an absence of association between and an empty one then? I feel like is isNull and isEmpty operators would clash in some way.

Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: @PatRolland - Not quite following

Comment (PatRolland) [reply elide-core/src/main/java/com/yahoo/elide/core/filter/Operator.java]: Doesn't matter I'm being picky 😄

File: elide-core/src/main/java/com/yahoo/elide/core/filter/IsEmptyPredicate.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ * See LICENSE file in project root for terms.
+ */
+package com.yahoo.elide.core.filter;
+
+import com.yahoo.elide.core.Path;
+
+import java.util.Collections;
+
+/**
+ * Is EMopty Predicate Class.
```
Comment (aklish) [inline elide-core/src/main/java/com/yahoo/elide/core/filter/IsEmptyPredicate.java]: typo

**Pull request merged**